controller/rats_manager.py

```diff
@@ -0,0 +1,191 @@
+import time
+import traceback
+from datetime import datetime
+from typing import Optional
+
+from misc import daemon
+from submodules.model import enums
+from misc.notification import (
+    send_notification_created,
+)
+from submodules.model import (
+    RecordAttributeTokenStatistics,
+)
+from submodules.model.enums import AttributeState
+from submodules.model.business_objects import (
+    project,
+    attribute,
+    record,
+    tokenization,
+    general,
+    notification,
+)
+from handler.tokenizer_handler import get_tokenizer_by_project
+from misc.util import get_docs_from_db, send_websocket_update
+from submodules.model.models import RecordTokenizationTask
+
+
+def trigger_rats_creation(
+    project_id: str, user_id: str, attribute_id: Optional[str] = None
+) -> None:
+    initial_count = record.count_missing_rats_records(project_id, attribute_id)
+    task = tokenization.create_tokenization_task(
+        project_id,
+        user_id,
+        enums.TokenizerTask.TYPE_TOKEN_STATISTICS.value,
+        with_commit=True,
+    )
+    daemon.run(
+        create_rats_entries,
+        project_id,
+        user_id,
+        attribute_id,
+        str(task.id),
+        initial_count,
+    )
+
+
+# rats = record_attribute_token_statistics
+def create_rats_entries(
+    project_id: str,
+    user_id: str,
+    attribute_id: str,
+    task_id: str,
+    initial_count: int,
+) -> None:
+    session_token = general.get_ctx_token()
+    if not project.get(project_id):
+        # project was deleted in the meantime
+        return
+    try:
+        tokenization_task = __set_up_statistic_calculation(
+            project_id, task_id, initial_count
+        )
+        i = 0
+        while initial_count > record.count_tokenized_records(project_id):
+            if i > 9:
+                print("Docbins missing", flush=True)
+                raise Exception("Docbins missing")
+            time.sleep(1)
+            i += 1
+        if attribute_id:
+            text_attribute = attribute.get(project_id, attribute_id)
+            text_attributes = {text_attribute.name: text_attribute.id}
+        else:
+            text_attributes = attribute.get_text_attributes(
+                project_id,
+                state_filter=[
+                    AttributeState.UPLOADED.value,
+                    AttributeState.USABLE.value,
+                    AttributeState.RUNNING.value,
+                ],
+            )
+        vocab = get_tokenizer_by_project(project_id).vocab
+        record_set = record.get_missing_rats_records(project_id, 100, attribute_id)
+        chunk = 0
+        while record_set:
+            entries = []
+            for record_item in record_set:
+                docs = get_docs_from_db(project_id, str(record_item.record_id), vocab)
```

Comment:
this is cold code however i think that we might still want to change it. Currently we select a record_set (100 records) then for each of these records we collect the byte data and count etc. Might be better to also collect from the db in chuncks instead of for each

Reply:
tbh i am not really sure what cold code means and the first few google results werent informative either. The second part about collecting it in chunks is valid and can be done, would discuss this

Reply:
typo it's "old" code :D

Reply:
changed query so that docs are read with records together